transportmiddleware/interface_test.go

```diff
@@ -0,0 +1,85 @@
+package transportmiddleware
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/yarpc/yab/transport"
+)
+
+type headerTransportMiddleware struct {
+	wantErr bool
+}
+
+func (tm headerTransportMiddleware) Apply(ctx context.Context, req *transport.Request) (*transport.Request, error) {
+	req.Headers["foo"] = "bar"
+	if tm.wantErr {
+		return nil, errors.New("bad apply")
+	}
+	return req, nil
+}
+
+func TestTransportMiddleware(t *testing.T) {
+	tests := []struct {
+		dontRegister bool
+		wantErr      bool
+		testRace     bool
+	}{
+		{ /* run without options */ },
+		{wantErr: true},
+		{dontRegister: true},
+		{testRace: true},
+	}
+	for idx, tt := range tests {
+		var restore func() = func() {}
+
+		// create the test middleware
+		tm := &headerTransportMiddleware{
+			wantErr: tt.wantErr,
+		}
+		if tt.testRace {
+			go Register(tm)
+		}
```

Comment:
This seems like it's testing a bunch of unrelated things together.

For the data race test at least, I would create a separate test.

You'll want multiple goroutines trying to Register at the same time. You
can use a `chan` to synchronize the Register call to increase likelihood
of conflict. You can do the same for the `restore` call as well.

```go
func TestRegisterRace(t *testing.T) {
  registerCh := make(chan struct{})
  restoreCh := make(chan struct{})
  var wg sync.WaitGroup

  for i := 0; i < 100; i++ {
    wg.Add(1)
    go func() {
      <-registerCh  // wait until register signal is received
      restore := Register(..)
      
      <- restoreCh
      restore()
      wg.Done()
    }()
  }

  close(registerCh)
  close(restoreCh)
  wg.Wait()
  
  // check that middleware is nil now
}
```

Reply:
This is great, haven't seen `close()` used like that before.. I like it 👍 

Also I don't think we can guarantee that the `registeredMiddleware` will be `nil` at the end, since it sort of implies we are doing many `register(a) -> restore(a) -> register(b) -> restore(b)` calls in a linear fashion. If we were to ever do `register(a) -> register(b) -> restore(a) -> restore(b)`, we would be left with `a` at the end, since that was the prior value when `b` was registered.

Reply:
Oh, true. Okay, yeah we can skip that test, I suppose.